packages/features/ee/round-robin/roundRobinReassignment.ts

```diff
@@ -0,0 +1,364 @@
+import EventManager from "@calcom/core/EventManager";
+import dayjs from "@calcom/dayjs";
+import { sendRoundRobinCancelledEmails, sendRoundRobinScheduledEmails } from "@calcom/emails";
+import { ensureAvailableUsers, getEventTypesFromDB } from "@calcom/features/bookings/lib/handleNewBooking";
+import type { User } from "@calcom/features/bookings/lib/handleNewBooking";
+import { scheduleEmailReminder } from "@calcom/features/ee/workflows/lib/reminders/emailReminderManager";
+import logger from "@calcom/lib/logger";
+import { getLuckyUser } from "@calcom/lib/server";
+import { getTranslation } from "@calcom/lib/server/i18n";
+import { getTimeFormatStringFromUserTimeFormat } from "@calcom/lib/timeFormat";
+import { prisma } from "@calcom/prisma";
+import { WorkflowActions, WorkflowTriggerEvents } from "@calcom/prisma/enums";
+import type { CalendarEvent } from "@calcom/types/Calendar";
+
+export const roundRobinReassignment = async ({
+  eventTypeId,
+  bookingId,
+}: {
+  eventTypeId: number;
+  bookingId: number;
+}) => {
+  const roundRobinReassignLogger = logger.getSubLogger({
+    prefix: ["roundRobinReassign", `${bookingId}`],
+  });
+
+  const eventType = await getEventTypesFromDB(eventTypeId);
+
+  if (!eventType) {
+    console.error(`Event type ${eventTypeId} not found`);
+    throw new Error("Event type not found");
+  }
+
+  eventType.users = eventType.hosts.map((host) => ({ ...host.user }));
+
+  const roundRobinHosts = eventType.hosts.filter((host) => !host.isFixed);
+
+  let booking = await prisma.booking.findUnique({
+    where: {
+      id: bookingId,
+    },
+    select: {
+      uid: true,
+      title: true,
+      startTime: true,
+      endTime: true,
+      userId: true,
+      customInputs: true,
+      responses: true,
+      destinationCalendar: true,
+      user: {
+        include: {
+          destinationCalendar: true,
+        },
+      },
+      attendees: {
+        select: {
+          id: true,
+          email: true,
+          name: true,
+          timeZone: true,
+          locale: true,
+        },
+      },
+      references: {
+        select: {
+          credentialId: true,
+        },
+      },
+    },
+  });
+
+  if (!booking) {
+    console.error(`Booking ${bookingId} not found`);
+    throw new Error("Booking not found");
+  }
+
+  if (!booking?.user) {
+    console.error(`No user associated with booking ${bookingId}`);
+    throw new Error("Booking not found");
+  }
+
+  const originalOrganizer = booking.user;
+
+  const attendeeEmailsSet = new Set(booking.attendees.map((attendee) => attendee.email));
```

Comment:
This set contains emails of both team members and attendees. We create a set because we're comparing team members to attendees in a few places. Rather than iterating through both arrays with a complexity of O(n * m) with a set we reduce it to O(n + m).